security/src/main/java/io/airlift/security/utils/KerberosUtil.java

```diff
@@ -0,0 +1,243 @@
+
+    static {
+        try {
+            SPNEGO_OID = new Oid("1.3.6.1.5.5.2");
+            KERBEROS_OID = new Oid("1.2.840.113554.1.2.2");
+            SUPPORTED_OIDS = new Oid[]{SPNEGO_OID, KERBEROS_OID};
+        } catch (GSSException e) {
+            throw new AuthenticationException(e);
+        }
+    }
+
+    /* Return the Kerberos login module name */
+    public static String getKrb5LoginModuleName()
+    {
+        return System.getProperty("java.vendor").contains("IBM")
+                ? "com.ibm.security.auth.module.Krb5LoginModule"
+                : "com.sun.security.auth.module.Krb5LoginModule";
+    }
+
+    public static Oid getOidInstance(String oidName)
+            throws GSSException, ClassNotFoundException, NoSuchFieldException,
+            IllegalAccessException
+    {
+        Class<?> oidClass;
+        if (System.getProperty("java.vendor").contains("IBM")) {
+            oidClass = Class.forName("com.ibm.security.jgss.GSSUtil");
+        }
+        else {
+            oidClass = Class.forName("sun.security.jgss.GSSUtil");
+        }
+        Field oidField = oidClass.getDeclaredField(oidName);
+        return (Oid) oidField.get(oidClass);
+    }
+
+    public static String getDefaultRealm()
+            throws ClassNotFoundException, NoSuchMethodException,
+            IllegalArgumentException, IllegalAccessException,
+            InvocationTargetException
+    {
+        Object kerbConf;
+        Class<?> classRef;
+        Method getInstanceMethod;
+        Method getDefaultRealmMethod;
+        if (System.getProperty("java.vendor").contains("IBM")) {
+            classRef = Class.forName("com.ibm.security.krb5.internal.Config");
+        }
+        else {
+            classRef = Class.forName("sun.security.krb5.Config");
+        }
+        getInstanceMethod = classRef.getMethod("getInstance", new Class[0]);
+        kerbConf = getInstanceMethod.invoke(classRef, new Object[0]);
+        getDefaultRealmMethod = classRef.getDeclaredMethod(
+                "getDefaultRealm",
+                new Class[0]);
+        return (String) getDefaultRealmMethod.invoke(kerbConf, new Object[0]);
+    }
+
+    /**
+     * Create Kerberos principal for a given service and hostname. It converts
+     * hostname to lower case. If hostname is null or "0.0.0.0", it uses
+     * dynamically looked-up fqdn of the current host instead.
+     *
+     * @param serviceName Service for which you want to generate the principal.
+     * @param serviceHostname Fully-qualified domain name.
+     * @return Converted Kerberos principal name.
+     * @throws UnknownHostException If no IP address for the local host could be found.
+     */
+    public static final String getServicePrincipal(String serviceName, String serviceHostname)
+            throws UnknownHostException
+    {
+        String fqdn = serviceHostname;
+        if (null == fqdn || fqdn.equals("") || fqdn.equals("0.0.0.0") || fqdn.equals("localhost")) {
```

Comment:
Two cases:
1) Client side: it depends on how people specify the service Uri and it may be out of our control.
2) Server side: we don't need to specify the host name and use 'null' to call this method.

Good point on IPv6. It looks like we followed a bad example: https://github.com/gsingers/kerberos/blob/master/src/main/java/com/lucidworks/security/authentication/util/KerberosUtil.java#L99.

Will update it to include IPv6 cases.